File: src/data/courses/dsa/string-hashing-problems/string-rolling-hash/problems-on-rolling-hash/code/java/ProblemsOnRollingHash.java
// Applications of Rolling Hash - Complete Implementations

import java.util.*;

class Solution {
    // Application 1: Pattern Matching (Rabin-Karp)
    public int rabinKarp(String text, String pattern) {
        int n = text.length(), m = pattern.length();
        long p = 31, mod = (long)1e9+7;
        
        long patHash = 0, txtHash = 0, powVal = 1;
        for (int i = 0; i < m; i++) {
            patHash = (patHash + (pattern.charAt(i)-'a'+1) * powVal) % mod;
            txtHash = (txtHash + (text.charAt(i)-'a'+1) * powVal) % mod;
            if (i < m-1) powVal = (powVal * p) % mod;
        }
        
        for (int i = 0; i <= n-m; i++) {
            if (patHash == txtHash && text.substring(i, i+m).equals(pattern))
                return i;
            if (i < n-m) {
                txtHash = (txtHash - (text.charAt(i)-'a'+1) + mod) % mod;
                txtHash = (txtHash / p) % mod;
                txtHash = (txtHash + (text.charAt(i+m)-'a'+1) * powVal) % mod;
            }
        }
        return -1;
    }
    
    // Application 2: Duplicate Substring Detection
    public List<Integer> findDuplicates(String s, int k) {
        int n = s.length();
        Map<Long, List<Integer>> hashMap = new HashMap<>();
        long p = 31, mod = (long)1e9+7;
        
        long hash = 0, powVal = 1;
        for (int i = 0; i < k; i++) {
            hash = (hash + (s.charAt(i)-'a'+1) * powVal) % mod;
            if (i < k-1) powVal = (powVal * p) % mod;
        }
        hashMap.computeIfAbsent(hash, x -> new ArrayList<>()).add(0);
        
        for (int i = k; i < n; i++) {
            hash = (hash - (s.charAt(i-k)-'a'+1) + mod) % mod;
            hash = (hash / p) % mod;
            hash = (hash + (s.charAt(i)-'a'+1) * powVal) % mod;
            hashMap.computeIfAbsent(hash, x -> new ArrayList<>()).add(i-k+1);
        }
        
        List<Integer> result = new ArrayList<>();
        for (List<Integer> indices : hashMap.values()) {
            if (indices.size() > 1) {
                result.addAll(indices);
            }
        }
        return result;
    }
    
    // Application 3: Longest Common Substring
    public int longestCommonSubstring(String s1, String s2) {
        int left = 0, right = Math.min(s1.length(), s2.length());
        int result = 0;
        
        while (left <= right) {
            int mid = (left + right) / 2;
            if (hasCommon(s1, s2, mid)) {
                result = mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return result;
    }
    
    private boolean hasCommon(String s1, String s2, int len) {
        Set<Long> hashes = new HashSet<>();
        long p = 31, mod = (long)1e9+7;
        
        long hash = 0, powVal = 1;
        for (int i = 0; i < len; i++) {
            hash = (hash * p + (s1.charAt(i)-'a'+1)) % mod;
            if (i < len-1) powVal = (powVal * p) % mod;
        }
        hashes.add(hash);
        
        for (int i = len; i < s1.length(); i++) {
            hash = (hash - (s1.charAt(i-len)-'a'+1) * powVal % mod + mod) % mod;
            hash = (hash * p + (s1.charAt(i)-'a'+1)) % mod;
            hashes.add(hash);
        }
        
        hash = 0;
        for (int i = 0; i < len; i++) {
            hash = (hash * p + (s2.charAt(i)-'a'+1)) % mod;
        }
        if (hashes.contains(hash)) return true;
        
        for (int i = len; i < s2.length(); i++) {
            hash = (hash - (s2.charAt(i-len)-'a'+1) * powVal % mod + mod) % mod;
            hash = (hash * p + (s2.charAt(i)-'a'+1)) % mod;
            if (hashes.contains(hash)) return true;
        }
        return false;
    }
    
    // Application 4: Longest Repeated Substring
    public int longestRepeatedSubstring(String s) {
        int left = 1, right = s.length();
        int result = 0;
        
        while (left <= right) {
            int mid = (left + right) / 2;
            if (hasRepeated(s, mid)) {
                result = mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return result;
    }
    
    private boolean hasRepeated(String s, int length) {
        Map<Long, Integer> seen = new HashMap<>();
        long p = 31, mod = (long)1e9+7;
        
        long hash = 0, powVal = 1;
        for (int i = 0; i < length; i++) {
            hash = (hash * p + (s.charAt(i)-'a'+1)) % mod;
            if (i < length-1) powVal = (powVal * p) % mod;
        }
        seen.put(hash, 0);
        
        for (int i = length; i < s.length(); i++) {
            hash = (hash - (s.charAt(i-length)-'a'+1) * powVal % mod + mod) % mod;
            hash = (hash * p + (s.charAt(i)-'a'+1)) % mod;
            if (seen.containsKey(hash)) return true;
            seen.put(hash, i - length + 1);
        }
        return false;
    }
}

File: src/data/courses/dsa/string-hashing-problems/string-rolling-hash/problems-on-rolling-hash/code/python/problems-on-rolling-hash.py
# Applications of Rolling Hash - Complete Implementations

# Application 1: Pattern Matching (Rabin-Karp)
def rabin_karp(text, pattern):
    n, m = len(text), len(pattern)
    p, mod = 31, 10**9+7
    
    pat_hash = txt_hash = pow_val = 0
    for i in range(m):
        pat_hash = (pat_hash + (ord(pattern[i])-ord('a')+1) * (p**i)) % mod
        txt_hash = (txt_hash + (ord(text[i])-ord('a')+1) * (p**i)) % mod
    pow_val = p**(m-1)
    
    for i in range(n-m+1):
        if pat_hash == txt_hash and text[i:i+m] == pattern:
            return i
        if i < n-m:
            txt_hash = (txt_hash - (ord(text[i])-ord('a')+1)) // p
            txt_hash = (txt_hash + (ord(text[i+m])-ord('a')+1) * pow_val) % mod
    return -1

# Application 2: Duplicate Substring Detection
def find_duplicates(s, k):
    n = len(s)
    hash_map = {}
    p, mod = 31, 10**9+7
    
    hash_val = pow_val = 0
    for i in range(k):
        hash_val = (hash_val + (ord(s[i])-ord('a')+1) * (p**i)) % mod
    pow_val = p**(k-1)
    hash_map[hash_val] = [0]
    
    for i in range(k, n):
        hash_val = (hash_val - (ord(s[i-k])-ord('a')+1)) // p
        hash_val = (hash_val + (ord(s[i])-ord('a')+1) * pow_val) % mod
        hash_map.setdefault(hash_val, []).append(i-k+1)
    
    result = []
    for indices in hash_map.values():
        if len(indices) > 1:
            result.extend(indices)
    return result

# Application 3: Longest Common Substring
def longest_common_substring(s1, s2):
    def has_common(length):
        hashes = set()
        p, mod = 31, 10**9+7
        
        hash_val = 0
        for i in range(length):
            hash_val = (hash_val * p + ord(s1[i])-ord('a')+1) % mod
        hashes.add(hash_val)
        
        for i in range(length, len(s1)):
            hash_val = (hash_val * p + ord(s1[i])-ord('a')+1) % mod
            hash_val = (hash_val - (ord(s1[i-length])-ord('a')+1) * pow(p, length, mod)) % mod
            hashes.add(hash_val)
        
        hash_val = 0
        for i in range(length):
            hash_val = (hash_val * p + ord(s2[i])-ord('a')+1) % mod
        if hash_val in hashes:
            return True
        
        for i in range(length, len(s2)):
            hash_val = (hash_val * p + ord(s2[i])-ord('a')+1) % mod
            hash_val = (hash_val - (ord(s2[i-length])-ord('a')+1) * pow(p, length, mod)) % mod
            if hash_val in hashes:
                return True
        return False
    
    left, right = 0, min(len(s1), len(s2))
    result = 0
    
    while left <= right:
        mid = (left + right) // 2
        if has_common(mid):
            result = mid
            left = mid + 1
        else:
            right = mid - 1
    return result

# Application 4: Longest Repeated Substring
def longest_repeated_substring(s):
    def has_repeated(length):
        seen = {}
        p, mod = 31, 10**9+7
        
        hash_val = 0
        for i in range(length):
            hash_val = (hash_val * p + ord(s[i])-ord('a')+1) % mod
        seen[hash_val] = 0
        
        for i in range(length, len(s)):
            hash_val = (hash_val * p + ord(s[i])-ord('a')+1) % mod
            hash_val = (hash_val - (ord(s[i-length])-ord('a')+1) * pow(p, length, mod)) % mod
            if hash_val in seen:
                return True
            seen[hash_val] = i - length + 1
        return False
    
    left, right = 1, len(s)
    result = 0
    
    while left <= right:
        mid = (left + right) // 2
        if has_repeated(mid):
            result = mid
            left = mid + 1
        else:
            right = mid - 1
    return result
